chainer/functions/connection/multihead_attention.py

```diff
@@ -0,0 +1,298 @@
+import typing as tp  # NOQA
+
+import chainer
+from chainer.functions.activation import softmax
+from chainer.functions.array import concat
+from chainer.functions.array import expand_dims
+from chainer.functions.array import split_axis
+from chainer.functions.array import repeat
+from chainer.functions.array import reshape
+from chainer.functions.array import transpose
+from chainer.functions.array import where
+from chainer.functions.math import average
+from chainer.functions.math import matmul
+from chainer.functions.noise import dropout
+from chainer.functions.connection import linear
+from chainer import types
+from chainer import variable
+
+
+InputType = tp.Union[variable.Variable, types.NdArray]
+
+
+def generate_zeros(device, shape, dtype):
+    with chainer.using_device(device):
+        return device.xp.zeros(shape=shape, dtype=dtype)
+
+
+def generate_ones(device, shape, dtype):
+    with chainer.using_device(device):
+        return device.xp.ones(shape=shape, dtype=dtype)
+
+
+def multihead_attention(
+    n_head,                   # type: int
+    embedding_size,           # type: int
+    query,                    # type: InputType
+    key,                      # type: InputType
+    value,                    # type: InputType
+    proj_in_W,                # type: tp.Union[variable.Variable, tp.Tuple[variable.Variable, variable.Variable, variable.Variable]]  # NOQA
+    proj_in_b,                # type: tp.Optional[variable.Variable]
+    bias_k,                   # type: variable.Variable
+    bias_v,                   # type: variable.Variable
+    proj_out_W,               # type: variable.Variable
+    proj_out_b,               # type: variable.Variable
+    add_zero_attn=False,      # type: bool
+    attention_dropout=0,      # type: float
+    post_dropout=0,           # type: float
+    key_padding_mask=None,    # type: tp.Optional[InputType]
+    attn_mask=None,           # type: tp.Optional[InputType]
+    dot_product_scaler=None,  # type: tp.Optional[float]
+    softmax_scaler=1.0,       # type: float
+    return_weights=True       # type: bool
```

Comment:
Are you planning to implement `static_k` and `static_v`, which are implemented in Pytorch?
While their name is a little confusing, they are the values computed in here:
https://github.com/chainer/chainer/blob/723fdd97b47e8848569dfa59f0ff655eaad1c5a1/chainer/functions/connection/multihead_attention.py#L203-L206

If we use the same k and v many times, we can reuse the values without recomputing. The static_k and _v are prepared for the purpose. For example, a typical transformer for machine translation encodes a source sentence via self-attention and outputs sequential vectors, and in its decoder, (in addition to self-attention) it performs attention over the vectors in each layer. In this case, the decoder attends the same set of vectors (for k and v) many times.

In my opinion, these features are just additional. And, we can omit it in this first PR.

(cf. In pytorch and its original source, fairseq, the arguments are used in these lines.)
https://github.com/pytorch/pytorch/blob/883628cb5c31c89323cef52e96310c92abecfcbe/torch/nn/functional.py#L3318-L3326
https://github.com/pytorch/fairseq/blob/3e3fe72299980f53262880e24e372ed7d785093c/fairseq/modules/multihead_attention.py#L180

Reply:
Thank you for your clarification of `static_k` and `static_v`.

IIRC, I tried to implement those `static` first, but I gave it up to avoid making this PR too complicated. Also, I thought it reasonable to create a follow-up PR of improved unit tests and `static_k` and `static_v`.

What do you think?

Reply:
I agree with you. `static_` stuff would make this PR complicated. It seems good to skip now.